text/0000-subslice-pattern-syntax.md

```diff
@@ -0,0 +1,207 @@
+- Feature Name: slice_patterns
+- Start Date: 2018-03-08
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Permit matching sub-slices and sub-arrays with the syntax `..`.  
+Binding a variable to the expression matched by a subslice pattern can be done
+using syntax `<IDENT> @ ..` similar to the existing `<IDENT> @ <PAT>` syntax, for example:
+
+```rust
+// Binding a sub-array:
+let [x, y @ .., z] = [1, 2, 3, 4]; // `y: [i32, 2] = [2, 3]`
+
+// Binding a sub-slice:
+let [x, y @ .., z]: &[u8] = &[1, 2, 3, 4]; // `y: &[i32] = &[2, 3]`
+```
+
+# Motivation
+[motivation]: #motivation
+
+## General motivation
+Stabilization of slice pattern with subslices is currently blocked on finalizing syntax for
+these subslices.  
+This RFC proposes a syntax for stabilization.
+
+## Motivation for the specific syntax
+
+### The shortcut form: `..`
+
+This form is already used in the meaning "rest of the list" in struct patterns, tuple struct
+patterns and tuple patterns so it would be logical to use it for slice patterns as well.  
+And indeed, in unstable Rust `..` is used in this meaning since long before 1.0.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Sub-slices and sub-arrays can be matched using `..` and `<IDENT> @ ..` can be used to bind
+these sub-slices and sub-arrays to an identifier.
+
+```rust
+// Matching slices using `ref` patterns:
+let v = vec![1, 2, 3];
+match v[..] {
+    [1, ref subslice @ .., 4] => assert_eq!(subslice.len(), 1),
+    [5, ref subslice @ ..] => assert_eq!(subslice.len(), 2),
+    [ref subslice @ .., 6] => assert_eq!(subslice.len(), 2),
+    [x, .., y] => assert!(v.len() >= 2),
+    [..] => {} // Always matches
+}
+
+// Matching slices using default-binding-modes:
+let v = vec![1, 2, 3];
+match &v[..] {
+    [1, subslice @ .., 4] => assert_eq!(subslice.len(), 1),
+    [5, subslice @ ..] => assert_eq!(subslice.len(), 2),
+    [subslice @ .., 6] => assert_eq!(subslice.len(), 2),
+    [x, .., y] => assert!(v.len() >= 2),
+    [..] => {} // Always matches
+}
+
+// Matching arrays by-value:
+let v = [1, 2, 3];
+match v {
+  [1, subarray @ .., 3] => assert_eq!(subarray, [2]),
+  [5, subarray @ ..] => has_type::<[i32; 2]>(subarray),
+  [subarray @ .., 6] => has_type::<[i32, 2]>(subarray),
+  [x, .., y] => has_type::<[i32, 1]>(x),
+  [..] => {},
+}
+```
```

Comment:
And what about `let v = [1, 2, 3]; match &mut? v { ... }`? -- It would be good to state whether that is allowed and what types the bindings have (e.g. `has_type::<&mut? [i32; 2]>(subarray)` and such).